File: DSA/3. Expensive Expeditions/Part 2 (Medium)/Python/judge.py
from collections import Counter
from solution import Solution

class Judge:
    def run_tests(self):
        test_cases = [
            "Hh-e-ell-l-lo-o",
            "Goodbye-------",
            "Peek----a-Boo",
        ]
        for test_case in test_cases:
            self.test(test_case, self.test_case_solution_output(test_case))

    def test(self, p, expected):
        solution = Solution()
        result = solution.decode_message(p)
        passed = result == expected
        print(f"Message: {p}")
        print(f"Expected: {expected}")
        print(f"Your Output: {result}")
        print("Test " + ("PASSED" if passed else "FAILED"))
        print("-----------")

    def test_case_solution_output(self, m):
        stack = []
        for char in m:
            if char == '-':
                stack.pop()
            else:
                stack.append(char)
        return ''.join(stack)

File: DSA/3. Expensive Expeditions/Part 2 (Medium)/Java/Judge.java
import java.util.*;

public class Judge {

    public void runTests() {
        String[] testCases = {
            "Hh-e-ell-l-lo-o",
            "Goodbye-------",
            "Peek----a-Boo",
        };

        runTests(testCases);
    }

    public void runTests(String[] testCases) {
        for(String testCase : testCases) {
            test(testCase, testCaseSolutionOutput(testCase));
        }
    }

    private boolean test(String m, String expected) {
        Solution sol = new Solution();
        String result = sol.decodeMessage(m);
        boolean passed = result.equals(expected);
        System.out.println("Message: " + m);
        System.out.println("Expected: " + expected);
        System.out.println("Your Output: " + result);
        System.out.println(passed ? "Test PASSED" : "Test FAILED");
        System.out.println("-----------");
        return passed;
    }

    private String testCaseSolutionOutput(String m) {
        Stack<Character> stack = new Stack<>();
        for (char c : m.toCharArray()) {
            if (c == '-') {
                if (!stack.isEmpty()) {
                    stack.pop();
                }
            } else {
                stack.push(c);
            }
        }
        StringBuilder result = new StringBuilder();
        for (char c : stack) {
            result.append(c);
        }
        return result.toString();
    }

}
